homeassistant/helpers/script.py

```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
     def run(self, variables=None, context=None):
         """Run script."""
         asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
+            self.async_run(variables, context), self._hass.loop
         ).result()
 
     async def async_run(
         self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
-
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
-            try:
-                await self._handle_action(action, variables, context)
-            except _SuspendScript:
-                # Store next step to take and notify change listeners
-                self._cur = cur + 1
-                if self._change_listener:
-                    self.hass.async_add_job(self._change_listener)
-                return
-            except _StopScript:
-                break
-            except Exception:
-                # Store the step that had an exception
-                self._exception_step = cur
-                # Set script to not running
-                self._cur = -1
-                self.last_action = None
-                # Pass exception on.
-                raise
-
-        # Set script to not-running.
-        self._cur = -1
-        self.last_action = None
-        if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+    ) -> bool:
+        """Run script."""
+        if self.is_running:
+            if self._if_running == IF_RUNNING_IGNORE:
+                self._log("Skipping script")
+                return False
+
+            if self._if_running == IF_RUNNING_ERROR:
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
 
-    def stop(self) -> None:
-        """Stop running script."""
-        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
+        self.last_triggered = utcnow()
+        if self._run_mode == RUN_MODE_LEGACY:
+            if self.is_running:
```

Comment:
Feels a little weird that you're checking `self.is_running`, I know it's just a wrapped, but you could also do:
```suggestion
            if self._runs:
```

Reply:
Yeah, that's probably a little clearer. Will do.